### DISCUSSION TITLE
Compile time only attribute for source generators

### DISCUSSION THREAD

**Discussion by thargol1** [dotnet/csharplang#4732] [Language Ideas] [answered]
# Compile time only attribute for source generators

I've been experimenting with source generators and it works great. Only I'm not happy with how to trigger a source generator.

Most examples use an attribute to indicate where source should be generated.

In my experiments I also used comments:
```csharp
//# ValueStruct: double,json,comparable,plusminus
public partial struct Celcius
{
	private const string PostFix = "°C";
	public bool IsValid => Value >= -273.15d;
}
```
And the source generator than generates the code for a single value struct with json conversion, validation, + and - operators, comparers and string parsing.

Using comments is technically easy using `GetLeadingTrivia()` but the big disavantage is that you have to parse the comment yourself and there is no validation during edit time.

I don't want to use attributes because they get emitted in to the resulting binary which results in unnecessary bytes spend and leaks some implementation details.

Therefor I would like to propose attributes that are available during compile time to source generators but don't get emmitted in the resulting binary.

Possible syntaxes:
```csharp
[[MyAttribute]]  // using double brackets to indicate a non emitting attribute

public MyAttribute : NoEmitAttribute // using a special base class

[NoEmit] public MyAttribute : Attribute // using an attribute on the attribute
```

Every usage of the attribute in actual code, outside the source generator, should generate a compiler error, like when calling `GetCustomAttribute<MyAttribute>`.

Or perhaps a different implementation is possible like:

```csharp
sourcegenerator(sourcegeneratoridentifier, param1, param2, .... paramx)
{
	public partial struct Celcius
	{
		private const string PostFix = "°C";
		public bool IsValid => Value >= -273.15d;
	}
}
```
Where `sourcegenerator` is a new keyword and calls the specified source generator with the embedded syntaxtree and specified parameters.

**Comment by chrisoverzero** [comment]
> I don't want to use attributes because they get emitted in to the resulting binary […]

This has been discussed elsewhere (#4715), and [the solution](https://github.com/dotnet/csharplang/discussions/4715#discussioncomment-685026) is to apply `ConditionalAttribute` with a value of `"DEBUG"` to the custom attributes.

**Comment by ufcpp** [comment, marked as answer]
This was also [discussed in 2015](https://github.com/dotnet/csharplang/blob/main/meetings/2015/LDM-2015-09-08.md).

**Reply by thargol1** [reply]
Hmm, I have to work on my search skills, I did try to find and existing issue, but didn't find it.

Thanks for the tip with `ConditionalAttribute`.

This issue can be closed.
